# Cross-Validation with Ridge and Lasso

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 200)
#SK LEARN
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import train_test_split
import sklearn.metrics as metrics
from sklearn.metrics import mean_squared_error


We will euse the Kings County housing dataset in this notebook

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/learn-co-curriculum/dsc-mod-2-project-v2-1/master/kc_house_data.csv',index_col=0)

## Data Cleaning and Prep

In [3]:
df['date'] = pd.to_datetime(df['date'])

df['yr_sold']=  df['date'].map(lambda x: x.year)

df['yrs_old'] =  df['yr_built'].map(lambda x: 2016-x)

df['yr_since_reno'] =  df['yr_renovated'].map(lambda x: 2016-x if x > 0 else np.nan)

df['yrs_since_update'] = df.apply(lambda x: min(x['yrs_old'], x['yr_since_reno']), axis=1)


df['bedrooms']=df['bedrooms'].map(lambda x: x if x < 10 else 10)

df.replace('?', 0, inplace=True)

df['sqft_basement'] = pd.to_numeric(df['sqft_basement'])

df.fillna(0, inplace=True)

In [4]:
zip_df = pd.get_dummies(df['zipcode'], drop_first=True)


In [5]:
df.shape

(21597, 24)

In [6]:
zip_df.shape

(21597, 69)

In [7]:
target= np.log(df.price)

features = df.drop(columns=['date', 'price', 'lat', 'long', 'yr_built', 'yr_renovated', 'yr_since_reno', 'zipcode'])

In [8]:
polynomial_features_2= PolynomialFeatures(degree=2, include_bias=False)
features_poly = polynomial_features_2.fit_transform(features)
poly_columns = polynomial_features_2.get_feature_names(features.columns)

In [9]:
features_poly = pd.DataFrame(features_poly, columns=poly_columns)

In [10]:
zip_df.reset_index(drop=True, inplace=True)

In [11]:
features_poly.head()

bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  \
0       3.0       1.00       1180.0    5650.0     1.0         0.0   0.0   
1       3.0       2.25       2570.0    7242.0     2.0         0.0   0.0   
2       2.0       1.00        770.0   10000.0     1.0         0.0   0.0   
3       4.0       3.00       1960.0    5000.0     1.0         0.0   0.0   
4       3.0       2.00       1680.0    8080.0     1.0         0.0   0.0   

   condition  grade  sqft_above  sqft_basement  sqft_living15  sqft_lot15  \
0        3.0    7.0      1180.0            0.0         1340.0      5650.0   
1        3.0    7.0      2170.0          400.0         1690.0      7639.0   
2        3.0    6.0       770.0            0.0         2720.0      8062.0   
3        5.0    7.0      1050.0          910.0         1360.0      5000.0   
4        3.0    8.0      1680.0            0.0         1800.0      7503.0   

   yr_sold  yrs_old  yrs_since_update  bedrooms^2  bedrooms bathrooms  \
0   2014.0     61.0              61.0         9.0                3.00   
1   2014.0     65.0              25.0         9.0                6.75   
2   2015.0     83.0              83.0         4.0                2.00   
3   2014.0     51.0              51.0        16.0               12.00   
4   2015.0     29.0              29.0         9.0                6.00   

   bedrooms sqft_living  bedrooms sqft_lot  bedrooms floors  \
0                3540.0            16950.0              3.0   
1                7710.0            21726.0              6.0   
2                1540.0            20000.0              2.0   
3                7840.0            20000.0              4.0   
4                5040.0            24240.0              3.0   

   bedrooms waterfront  bedrooms view  bedrooms condition  bedrooms grade  \
0                  0.0            0.0                 9.0            21.0   
1                  0.0            0.0                 9.0            21.0   
2                  0.0            0.0                 6.0            12.0   
3                  0.0            0.0                20.0            28.0   
4                  0.0            0.0                 9.0            24.0   

   bedrooms sqft_above  bedrooms sqft_basement  bedrooms sqft_living15  \
0               3540.0                     0.0                  4020.0   
1               6510.0                  1200.0                  5070.0   
2               1540.0                     0.0                  5440.0   
3               4200.0                  3640.0                  5440.0   
4               5040.0                     0.0                  5400.0   

   bedrooms sqft_lot15  bedrooms yr_sold  bedrooms yrs_old  \
0              16950.0            6042.0             183.0   
1              22917.0            6042.0             195.0   
2              16124.0            4030.0             166.0   
3              20000.0            8056.0             204.0   
4              22509.0            6045.0              87.0   

   bedrooms yrs_since_update  bathrooms^2  bathrooms sqft_living  \
0                      183.0       1.0000                 1180.0   
1                       75.0       5.0625                 5782.5   
2                      166.0       1.0000                  770.0   
3                      204.0       9.0000                 5880.0   
4                       87.0       4.0000                 3360.0   

   bathrooms sqft_lot  bathrooms floors  bathrooms waterfront  bathrooms view  \
0              5650.0               1.0                   0.0             0.0   
1             16294.5               4.5                   0.0             0.0   
2             10000.0               1.0                   0.0             0.0   
3             15000.0               3.0                   0.0             0.0   
4             16160.0               2.0                   0.0             0.0   

   bathrooms condition  bathrooms grade  bathrooms sqft_above  \
0                 3.00       

In [14]:
features_ploy = pd.merge(features_poly, zip_df, left_index=True, right_index=True)

In [16]:
features_ploy

bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  \
0           3.0       1.00       1180.0    5650.0     1.0         0.0   0.0   
1           3.0       2.25       2570.0    7242.0     2.0         0.0   0.0   
2           2.0       1.00        770.0   10000.0     1.0         0.0   0.0   
3           4.0       3.00       1960.0    5000.0     1.0         0.0   0.0   
4           3.0       2.00       1680.0    8080.0     1.0         0.0   0.0   
...         ...        ...          ...       ...     ...         ...   ...   
21592       3.0       2.50       1530.0    1131.0     3.0         0.0   0.0   
21593       4.0       2.50       2310.0    5813.0     2.0         0.0   0.0   
21594       2.0       0.75       1020.0    1350.0     2.0         0.0   0.0   
21595       3.0       2.50       1600.0    2388.0     2.0         0.0   0.0   
21596       2.0       0.75       1020.0    1076.0     2.0         0.0   0.0   

       condition  grade  sqft_above  sqft_basement  sqft_living15  sqft_lot15  \
0            3.0    7.0      1180.0            0.0         1340.0      5650.0   
1            3.0    7.0      2170.0          400.0         1690.0      7639.0   
2            3.0    6.0       770.0            0.0         2720.0      8062.0   
3            5.0    7.0      1050.0          910.0         1360.0      5000.0   
4            3.0    8.0      1680.0            0.0         1800.0      7503.0   
...          ...    ...         ...            ...            ...         ...   
21592        3.0    8.0      1530.0            0.0         1530.0      1509.0   
21593        3.0    8.0      2310.0            0.0         1830.0      7200.0   
21594        3.0    7.0      1020.0            0.0         1020.0      2007.0   
21595        3.0    8.0      1600.0            0.0         1410.0      1287.0   
21596        3.0    7.0      1020.0            0.0         1020.0      1357.0   

       yr_sold  yrs_old  yrs_since_update  bedrooms^2  bedrooms bathrooms  \
0       2014.0     61.0              61.0         9.0                3.00   
1       2014.0     65.0              25.0         9.0                6.75   
2       2015.0     83.0              83.0         4.0                2.00   
3       2014.0     51.0              51.0        16.0               12.00   
4       2015.0     29.0              29.0         9.0                6.00   
...        ...      ...               ...         ...                 ...   
21592   2014.0      7.0               7.0         9.0                7.50   
21593   2015.0      2.0               2.0        16.0               10.00   
21594   2014.0      7.0               7.0         4.0                1.50   
21595   2015.0     12.0              12.0         9.0                7.50   
21596   2014.0      8.0               8.0         4.0                1.50   

       bedrooms sqft_living  bedrooms sqft_lot  bedrooms floors  \
0                    3540.0            16950.0              3.0   
1                    7710.0            21726.0              6.0   
2                    1540.0            20000.0              2.0   
3                    7840.0            20000.0              4.0   
4                    5040.0            24240.0              3.0   
...                     ...                ...              ...   
21592                4590.0             3393.0              9.0   
21593                9240.0            23252.0              8.0   
21594                2040.0             2700.0              4.0   
21595                4800.0             7164.0              6.0   
21596                2040.0             2152.0              4.0   

       bedrooms waterfront  bedrooms view  bedrooms condition  bedrooms grade  \
0                      0.0            0.0                 9.0            21.0   
1                      0.0            0.0                 9.0            21.0   
2                      0.0            0.0                 6.0            12.0   
3                      0.0     

In [17]:
#call train_test_split on the data and capture the results
X_train, X_test, y_train, y_test = train_test_split(features_poly, target, random_state=22,test_size=0.25)


In [18]:
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled =pd.DataFrame(data=scaler.transform(X_train), columns=poly_columns)
X_test_scaled =pd.DataFrame(data=scaler.transform(X_test), columns=poly_columns)

In [19]:
X_test_scaled

bedrooms  bathrooms  sqft_living  sqft_lot    floors  waterfront  \
0     0.698568  -0.474522     0.136010 -0.181476  0.930263   -0.078021   
1     0.698568  -0.474522     0.103180 -0.254240 -0.913867   -0.078021   
2    -0.411261  -0.474522    -0.115684 -0.116716 -0.913867   -0.078021   
3     0.698568  -0.474522    -0.389264 -0.153583 -0.913867   -0.078021   
4    -0.411261   0.176627    -0.630015 -0.181112 -0.913867   -0.078021   
...        ...        ...          ...       ...       ...         ...   
5395 -1.521091  -1.451244    -1.034914 -0.301779 -0.913867   -0.078021   
5396 -0.411261  -0.148948    -0.257946 -0.291228 -0.913867   -0.078021   
5397  1.808398   3.106794     2.105788 -0.202990  0.930263   -0.078021   
5398 -0.411261  -0.148948    -0.257946  0.042031 -0.913867   -0.078021   
5399 -0.411261  -1.451244    -1.188119 -0.144003  0.008198   -0.078021   

          view  condition     grade  sqft_above  sqft_basement  sqft_living15  \
0    -0.303802  -0.628011  0.291369    0.501745      -0.647824      -0.107805   
1    -0.303802  -0.628011 -0.563268   -0.625955       1.408665      -0.342119   
2    -0.303802   0.912485  0.291369    0.222851      -0.647824      -0.019937   
3    -0.303802  -0.628011 -0.563268   -1.123114       1.317266      -0.283540   
4    -0.303802  -0.628011 -0.563268   -0.771465       0.151922      -0.649656   
...        ...        ...       ...         ...            ...            ...   
5395 -0.303802   0.912485  0.291369   -0.795717      -0.647824      -0.444631   
5396 -0.303802   2.452981 -0.563268   -1.050359       1.454365      -0.737524   
5397 -0.303802  -0.628011  1.146006    2.684391      -0.647824       2.498938   
5398 -0.303802   2.452981 -0.563268   -1.001856       1.362965      -0.254251   
5399 -0.303802   0.912485 -1.417904   -0.965478      -0.647824      -0.503210   

      sqft_lot15   yr_sold   yrs_old  yrs_since_update  bedrooms^2  \
0      -0.190030  1.446652 -0.236968         -0.173623    0.572810   
1      -0.324746  1.446652  0.714699          0.791977    0.572810   
2      -0.160093 -0.691251  0.170889          0.240206   -0.474802   
3      -0.146996 -0.691251  0.000948          0.067777    0.572810   
4      -0.150776 -0.691251 -0.440897         -0.380538   -0.474802   
...          ...       ...       ...               ...         ...   
5395   -0.354683 -0.691251  1.496426          1.585149   -1.223096   
5396   -0.333203  1.446652  1.462438          1.550664   -0.474802   
5397   -0.227451  1.446652 -1.494529         -1.449595    1.919739   
5398    0.151698 -0.691251 -0.033040          0.033291   -0.474802   
5399   -0.163087  1.446652  1.088569          1.171321   -0.474802   

      bedrooms bathrooms  bedrooms sqft_living  bedrooms sqft_lot  \
0              -0.117102              0.269874          -0.146784   
1              -0.117102              0.245257          -0.227542   
2              -0.535811             -0.322988          -0.144025   
3              -0.117102             -0.123999          -0.115827   
4              -0.176917             -0.612238          -0.197628   
...                  ...                   ...                ...   
5395           -1.313414             -1.071758          -0.315838   
5396           -0.356364             -0.402993          -0.289289   
5397            3.591467              2.567470          -0.125484   
5398           -0.356364             -0.402993          -0.011884   
5399           -1.074152             -0.926107          -0.166738   

      bedrooms floors  bedrooms waterfront  bedrooms view  bedrooms condition  \
0            1.164248            -0.073901       -0.29044            0.128589   
1           -0.457905            -0.073901       -0.29044            0.128589   
2           -0.863443            -0.073901       -0.29044            0.128589   
3           -0.457905            -0.073901       -0.29044            0.128589   
4           -0.863443            -0.073901       -0.29044           -

In [20]:
#instantiate a linear regression object
lm = LinearRegression()

#fit the linear regression to the data
lm = lm.fit(X_train_scaled, y_train)


print(lm.intercept_)
print(lm.coef_)

13.049903861156425
[-2.91305445e+00  1.23870022e+01  2.07153924e+01 -4.39179096e+00
 -1.90808328e+01 -8.41455939e+00 -1.51481175e+01  3.36490873e+01
 -1.07452644e+01 -8.86740641e+00 -1.33832994e+01  9.00938543e+00
 -2.62928673e+00  1.48549963e-02  2.10498164e+01 -6.74514164e+01
  4.20955744e-02  3.60245391e-02 -1.77223612e-01  4.69499387e-02
  2.56430934e-02  5.35700275e-03 -1.93386422e-02 -2.32943924e-02
  3.72462724e-02  1.07697021e-01  7.81923727e-03  7.95898903e-02
 -1.07808292e-02  2.73691438e+00  4.69290332e-02  8.41215226e-02
 -8.72827541e-02  7.98901874e-01 -5.02274369e-03 -7.48220924e-02
  4.87133674e-03 -1.65172959e-02  3.24840845e-02  1.10047810e-01
 -5.89293144e-01 -2.08101398e-01 -8.55275875e-02 -2.37199242e-02
 -1.23101934e+01 -4.18824069e-02 -2.14053709e-02 -8.29762053e-01
  4.32568850e-02 -1.43949691e-01  4.00095127e-03 -7.88699499e-04
  4.87221244e-03 -2.03428149e-01  8.39757778e-01  6.52766937e-02
 -2.97020647e-01 -8.64609879e-02 -2.01017450e+01 -1.03290882e-01
  8.86

In [22]:
y_train_pred = lm.predict(X_train)
y_pred = lm.predict(X_test)

In [23]:
train_mae = metrics.mean_absolute_error(y_train, y_train_pred)
train_mse = metrics.mean_squared_error(y_train, y_train_pred)
train_rmse = np.sqrt(metrics.mean_squared_error(y_train, y_train_pred))


print('Mean Absolute Error:', train_mae )
print('Mean Squared Error:',  train_mse)
print('Root Mean Squared Error:' , train_rmse)

Mean Absolute Error: 188323862.07597286
Mean Squared Error: 8.28947623045237e+17
Root Mean Squared Error: 910465607.8321888


In [26]:
sum(np.abs(lm.coef_))

525.1409558976251

In [27]:
from sklearn.feature_selection import RFECV
ols = LinearRegression()

In [29]:
#recursive wrapper method

# Create recursive feature eliminator that scores features by mean squared errors
selector = RFECV(estimator=ols, step=3, cv=5, scoring='neg_mean_squared_error', verbose =1, n_jobs=-1)

# Fit recursive feature eliminator 
selector.fit(X_train_scaled, y_train)



Fitting estimator with 152 features.
Fitting estimator with 149 features.
Fitting estimator with 146 features.
Fitting estimator with 143 features.
Fitting estimator with 140 features.
Fitting estimator with 137 features.
Fitting estimator with 134 features.
Fitting estimator with 131 features.
Fitting estimator with 128 features.
Fitting estimator with 125 features.
Fitting estimator with 122 features.
Fitting estimator with 119 features.
Fitting estimator with 116 features.
Fitting estimator with 113 features.
Fitting estimator with 110 features.
Fitting estimator with 107 features.
Fitting estimator with 104 features.
Fitting estimator with 101 features.
Fitting estimator with 98 features.
Fitting estimator with 95 features.
Fitting estimator with 92 features.
Fitting estimator with 89 features.
Fitting estimator with 86 features.
Fitting estimator with 83 features.


RFECV(cv=5,
      estimator=LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
                                 normalize=False),
      min_features_to_select=1, n_jobs=-1, scoring='neg_mean_squared_error',
      step=3, verbose=1)

In [30]:
selected_columns = X_train_scaled.columns[selector.support_]
removed_columns = X_train_scaled.columns[~selector.support_]

In [31]:
len(selected_columns)

80

In [43]:
removed_columns

Index(['yr_sold', 'bedrooms sqft_lot', 'bedrooms floors',
       'bedrooms waterfront', 'bedrooms view', 'bedrooms condition',
       'bedrooms grade', 'bedrooms sqft_basement', 'bedrooms sqft_lot15',
       'bedrooms yrs_since_update', 'bathrooms sqft_lot',
       'bathrooms waterfront', 'bathrooms view', 'bathrooms condition',
       'bathrooms sqft_lot15', 'bathrooms yrs_since_update',
       'sqft_living waterfront', 'sqft_living view', 'sqft_living condition',
       'sqft_living sqft_lot15', 'sqft_living yrs_since_update', 'sqft_lot^2',
       'sqft_lot floors', 'sqft_lot waterfront', 'sqft_lot view',
       'sqft_lot condition', 'sqft_lot sqft_above', 'sqft_lot sqft_basement',
       'sqft_lot sqft_lot15', 'sqft_lot yrs_since_update', 'floors waterfront',
       'floors view', 'floors grade', 'floors sqft_lot15', 'floors yrs_old',
       'waterfront view', 'waterfront condition', 'waterfront grade',
       'waterfront sqft_above', 'waterfront sqft_basement',
       'waterfront s

In [32]:
lm_rfe = LinearRegression()

lm_rfe = lm_rfe.fit(X_train_scaled[selected_columns], y_train)

y_rfe=lm_rfe.predict(X_train_scaled[selected_columns])

trainRFE_rmse = np.sqrt(metrics.mean_squared_error(y_train, y_rfe))

print('Training Root Mean Squared Error:' , trainRFE_rmse)

y_pred_rfe = lm_rfe.predict(X_test_scaled[selected_columns])

testRFE_rmse = np.sqrt(metrics.mean_squared_error(y_test, y_pred_rfe))

print('Testing Root Mean Squared Error:' , testRFE_rmse)


print("Train RMSE", int(trainRFE_rmse), "Test RMSE: ", int(testRFE_rmse))

Training Root Mean Squared Error: 0.2934380808883874
Testing Root Mean Squared Error: 0.29886499473433853
Train RMSE 0 Test RMSE:  0


In [44]:
#Calculate sum of coefficients for this model
sum(np.abs(lm_rfe.coef_))

541.4110461177663

In [45]:
## training the model
from sklearn.linear_model import Lasso

lasso = Lasso(alpha=0.01, normalize=False)

lasso.fit(X_train,y_train)

y_train_pred_lasso = lasso.predict(X_train_scaled)
y_pred_lasso = lasso.predict(X_test_scaled)

train_rmse_lasso = metrics.mean_absolute_error(y_train, y_train_pred_lasso)
test_rmse_lasso = np.sqrt(metrics.mean_squared_error(y_test, y_pred_lasso))
print('Training Error: '+ str(train_rmse_lasso) )
print('Testing Error: '+ str(test_rmse_lasso) )



Training Error: 44.891729447054225
Testing Error: 44.88817795524521


/Users/flatironschool/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:475: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 700.174981176562, tolerance: 0.4424574981601789
  positive)


In [46]:
from sklearn.linear_model import SGDRegressor


In [47]:
lasso = SGDRegressor(penalty='l1', alpha=1, early_stopping=True, verbose=1)

lasso.fit(X_train_scaled,y_train)




-- Epoch 1
Norm: 6.83, NNZs: 0, Bias: 13.048590, T: 14577, Avg. loss: 0.968616
Total training time: 0.03 seconds.
-- Epoch 2
Norm: 6.91, NNZs: 0, Bias: 13.059233, T: 29154, Avg. loss: 0.121341
Total training time: 0.07 seconds.
-- Epoch 3
Norm: 6.97, NNZs: 0, Bias: 13.058774, T: 43731, Avg. loss: 0.121360
Total training time: 0.11 seconds.
-- Epoch 4
Norm: 7.02, NNZs: 0, Bias: 13.052976, T: 58308, Avg. loss: 0.121366
Total training time: 0.14 seconds.
-- Epoch 5
Norm: 7.07, NNZs: 0, Bias: 13.058428, T: 72885, Avg. loss: 0.121336
Total training time: 0.17 seconds.
-- Epoch 6
Norm: 7.11, NNZs: 0, Bias: 13.052514, T: 87462, Avg. loss: 0.121333
Total training time: 0.20 seconds.
Convergence after 6 epochs took 0.21 seconds


SGDRegressor(alpha=1, average=False, early_stopping=True, epsilon=0.1,
             eta0=0.01, fit_intercept=True, l1_ratio=0.15,
             learning_rate='invscaling', loss='squared_loss', max_iter=1000,
             n_iter_no_change=5, penalty='l1', power_t=0.25, random_state=None,
             shuffle=True, tol=0.001, validation_fraction=0.1, verbose=1,
             warm_start=False)

In [48]:
y_train_pred_lasso = lasso.predict(X_train_scaled)
y_pred_lasso = lasso.predict(X_test_scaled)

train_rmse_lasso = metrics.mean_absolute_error(y_train, y_train_pred_lasso)
test_rmse_lasso = np.sqrt(metrics.mean_squared_error(y_test, y_pred_lasso))
print('Training Error: '+ str(train_rmse_lasso) )
print('Testing Error: '+ str(test_rmse_lasso) )

Training Error: 0.41134272482995826
Testing Error: 0.5380729040193435


In [49]:
#Calculate sum of coefficients for this model

In [52]:
ridge = SGDRegressor(penalty='l2', alpha=1, early_stopping=True, verbose=1)

ridge.fit(X_train_scaled,y_train)



-- Epoch 1
Norm: 29963178409.69, NNZs: 152, Bias: -1240170486.549013, T: 14577, Avg. loss: 2941763450783918981120.000000
Total training time: 0.01 seconds.
-- Epoch 2
Norm: 35718231.52, NNZs: 152, Bias: 534755.295555, T: 29154, Avg. loss: 142626393679406628864.000000
Total training time: 0.04 seconds.
-- Epoch 3
Norm: 4064.99, NNZs: 152, Bias: 610.568530, T: 43731, Avg. loss: 43831842018299.242188
Total training time: 0.06 seconds.
-- Epoch 4
Norm: 1.14, NNZs: 152, Bias: 13.026520, T: 58308, Avg. loss: 1415791.847110
Total training time: 0.08 seconds.
-- Epoch 5
Norm: 0.11, NNZs: 152, Bias: 13.041625, T: 72885, Avg. loss: 0.072558
Total training time: 0.12 seconds.
-- Epoch 6
Norm: 0.11, NNZs: 152, Bias: 13.059753, T: 87462, Avg. loss: 0.054903
Total training time: 0.14 seconds.
-- Epoch 7
Norm: 0.18, NNZs: 152, Bias: 13.062134, T: 102039, Avg. loss: 0.055063
Total training time: 0.16 seconds.
-- Epoch 8
Norm: 0.10, NNZs: 152, Bias: 13.052131, T: 116616, Avg. loss: 0.049709
Total train

SGDRegressor(alpha=1, average=False, early_stopping=True, epsilon=0.1,
             eta0=0.01, fit_intercept=True, l1_ratio=0.15,
             learning_rate='invscaling', loss='squared_loss', max_iter=1000,
             n_iter_no_change=5, penalty='l2', power_t=0.25, random_state=None,
             shuffle=True, tol=0.001, validation_fraction=0.1, verbose=1,
             warm_start=False)

In [51]:
y_train_pred_ridge = ridge.predict(X_train_scaled)
y_pred_ridge = ridge.predict(X_test_scaled)

train_rmse_ridge = metrics.mean_absolute_error(y_train, y_train_pred_ridge)
test_rmse_ridge = np.sqrt(metrics.mean_squared_error(y_test, y_pred_ridge))
print('Training Error: '+ str(train_rmse_ridge) )
print('Testing Error: '+ str(test_rmse_ridge) )

Training Error: 8435285439.610875
Testing Error: 20584632107.038567


In [41]:
#Calculate sum of coefficients for this model

In [42]:
lasso_coef01 = pd.DataFrame(data=lasso.coef_).T
lasso_coef01.columns = X_train_scaled.columns
lasso_coef01 = lasso_coef01.T.sort_values(by=0).T
#lasso_coef01.plot(kind='bar', title='Modal Coefficients', legend=False, figsize=(16,8))
lasso_coef01.T

0
sqft_lot15^2                   -2.593066e+10
sqft_lot sqft_lot15            -1.786915e+10
sqft_lot waterfront            -1.184477e+10
view sqft_lot15                -1.124421e+10
sqft_above sqft_lot15          -9.042343e+09
floors sqft_lot15              -8.230898e+09
sqft_living sqft_lot15         -7.269173e+09
sqft_lot^2                     -7.176319e+09
sqft_lot view                  -6.110454e+09
bathrooms sqft_lot15           -5.700345e+09
sqft_lot yrs_since_update      -5.435658e+09
waterfront yrs_since_update    -4.801975e+09
sqft_above^2                   -2.563666e+09
bathrooms sqft_lot             -2.536380e+09
bedrooms sqft_living15         -2.429761e+09
view sqft_basement             -2.330407e+09
floors sqft_living15           -2.036117e+09
sqft_lot sqft_above            -1.967675e+09
view sqft_above                -1.828754e+09
sqft_basement sqft_living15    -1.786264e+09
view grade                     -1.708718e+09
bedrooms sqft_lot15            -1.660307e+09
bathrooms floors               -1.620632e+09
condition yrs_old              -1.613745e+09
sqft_living sqft_above         -1.585650e+09
condition sqft_basement        -1.563314e+09
grade yrs_old                  -1.541317e+09
sqft_lot grade                 -1.496408e+09
sqft_living view               -1.469081e+09
bathrooms^2                    -1.369697e+09
waterfront sqft_basement       -1.355825e+09
condition yrs_since_update     -1.343325e+09
bathrooms sqft_basement        -1.246877e+09
floors sqft_basement           -1.216032e+09
grade sqft_above               -1.167124e+09
bedrooms sqft_basement         -1.167078e+09
sqft_living15 yrs_old          -1.104631e+09
bedrooms yrs_since_update      -1.009621e+09
grade yrs_since_update         -9.122064e+08
sqft_living grade              -9.021165e+08
yr_sold^2                      -8.660925e+08
yr_sold                        -8.660925e+08
sqft_living condition          -6.497872e+08
sqft_living sqft_lot           -6.347360e+08
bathrooms                      -6.108565e+08
bathrooms yr_sold              -6.100985e+08
condition^2                    -5.998809e+08
sqft_living^2                  -5.970047e+08
bathrooms waterfront           -5.580733e+08
yrs_old^2                      -5.131454e+08
bathrooms sqft_living          -5.104949e+08
view^2                         -5.087546e+08
bedrooms condition             -4.214778e+08
bedrooms bathrooms             -3.441800e+08
floors sqft_above              -3.378794e+08
bathrooms yrs_old              -3.151211e+08
sqft_basement yrs_old          -3.052115e+08
condition grade                -2.455314e+08
bathrooms sqft_above           -2.395894e+08
floors condition               -2.267683e+08
sqft_above yr_sold             -1.926110e+08
sqft_above                     -1.901063e+08
sqft_living waterfront         -1.606571e+08
grade sqft_basement            -1.298341e+08
waterfront                     -1.059237e+08
waterfront^2                   -1.059237e+08
sqft_living15 yrs_since_update -8.608141e+07
waterfront yr_sold             -7.112620e+07
yrs_old                        -4.591831e+07
yr_sold yrs_old                -4.150282e+07
bathrooms grade                -3.474299e+07
grade^2                        -3.008567e+07
bedrooms sqft_living           -1.768809e+06
sqft_living yrs_since_update    3.021547e+06
sqft_living floors              1.380516e+07
view yrs_since_update           3.037602e+07
floors yr_sold                  3.539462e+07
floors                          4.073451e+07
sqft_basement yrs_since_update  8.197576e+07
yrs_since_update^2              9.282806e+07
sqft_basement                   9.752875e+07
sqft_basement yr_sold           1.004452e+08
view                            1.185842e+08
waterfront condition            1.208347e+08
view yr_sold                    1.247415e+08
bedrooms yrs_old                1.344074e+08
sqft_living yrs_old             1.483582e+08
condition sqft_above            1.617552e+08
sqft_above yrs_since_update     1.6953

We want to pick our best model, but this is more complicated than just choosing between linear regression, Lasso, or Ridge. We now have to also consider the different models that we get from different alpha values for Ridge and Lasso.


How do we determine the best model that will not overfit to the training data? 

___

## Cross Validation

Cross-validation is a statistical method used to protect against overfitting a predictive model, particularly in a case where the amount of data may be limited. In cross-validation, you make a fixed number of folds (or partitions) of the data, run the analysis on each fold, and then average the overall error estimate.

### Steps for K-fold cross-validation



1. Split the dataset into K **equal** partitions (or "folds").
2. Use fold 1 as the **testing set** and the union of the other folds as the **training set**.
3. Calculate **testing accuracy**.
4. Repeat steps 2 and 3 K times, using a **different fold** as the testing set each time.
5. Use the **average testing accuracy** as the estimate of out-of-sample accuracy.

Diagram of **5-fold cross-validation:**

<img src="https://miro.medium.com/max/1354/1*qPMFLEbvc8QQf38Cf77wQg.png">

In [53]:
# simulate splitting a dataset of 25 observations into 5 folds
from sklearn.model_selection import KFold
kf = KFold(n_splits=5, shuffle=False).split(range(25))

In [54]:
# print the contents of each training and testing set
print('{} {:^61} {}'.format('Iteration', 'Training set observations', 'Testing set observations'))
for iteration, data in enumerate(kf, start=1):
    print('{:^9} {} {:^25}'.format(iteration, data[0], str(data[1])))

Iteration                   Training set observations                   Testing set observations
    1     [ 5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24]        [0 1 2 3 4]       
    2     [ 0  1  2  3  4 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24]        [5 6 7 8 9]       
    3     [ 0  1  2  3  4  5  6  7  8  9 15 16 17 18 19 20 21 22 23 24]     [10 11 12 13 14]     
    4     [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 20 21 22 23 24]     [15 16 17 18 19]     
    5     [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]     [20 21 22 23 24]     


- Dataset contains **25 observations** (numbered 0 through 24)
- 5-fold cross-validation, thus it runs for **5 iterations**
- For each iteration, every observation is either in the training set or the testing set, **but not both**
- Every observation is in the testing set **exactly once**

### Comparing cross-validation to train/test split



Advantages of **cross-validation:**

- More accurate estimate of out-of-sample accuracy
- More "efficient" use of data (every observation is used for both training and testing)

Advantages of **train/test split:**

- Runs K times faster than K-fold cross-validation
- Simpler to examine the detailed results of the testing process

### Cross-validation recommendations



1. K can be any number, but **K=10** is generally recommended
2. For classification problems, **stratified sampling** is recommended for creating the folds
    - Each response class should be represented with equal proportions in each of the K folds
    - scikit-learn's `cross_val_score` function does this by default

In [55]:
from sklearn.linear_model import LassoCV, RidgeCV

In [56]:
lassoCV_model = LassoCV(cv=5, random_state=0, verbose=1)
lassoCV_model.fit(X_train_scaled, y_train)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    1.4s finished
/Users/flatironschool/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:475: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 9.018195076280335, tolerance: 0.4424574981601789
  positive)


LassoCV(alphas=None, copy_X=True, cv=5, eps=0.001, fit_intercept=True,
        max_iter=1000, n_alphas=100, n_jobs=None, normalize=False,
        positive=False, precompute='auto', random_state=0, selection='cyclic',
        tol=0.0001, verbose=1)

In [57]:
lassoCV_model.alpha_

0.0003732575008003002

In [58]:
RidgeCV_model = RidgeCV(cv=5)
RidgeCV_model.fit(X_train_scaled, y_train)

RidgeCV(alphas=array([ 0.1,  1. , 10. ]), cv=5, fit_intercept=True,
        gcv_mode=None, normalize=False, scoring=None, store_cv_values=False)

In [59]:
RidgeCV_model.alpha_

10.0

Now that we have used cross validation to help us determine the best **alpha** for Ridge and Lasso, we can then use those fitted models to compare on our test set.  

## Improvements to cross-validation

**Repeated cross-validation**

- Repeat cross-validation multiple times (with **different random splits** of the data) and average the results
- More reliable estimate of out-of-sample performance by **reducing the variance** associated with a single trial of cross-validation

**Creating a hold-out set**

- "Hold out" a portion of the data **before** beginning the model building process
- Locate the best model using cross-validation on the remaining data, and test it **using the hold-out set**
- More reliable estimate of out-of-sample performance since hold-out set is **truly out-of-sample**

**Feature engineering and selection within cross-validation iterations**

- Normally, feature engineering and selection occurs **before** cross-validation
- Instead, perform all feature engineering and selection **within each cross-validation iteration**
- More reliable estimate of out-of-sample performance since it **better mimics** the application of the model to out-of-sample data


## Resources


- scikit-learn documentation: [Cross-validation](http://scikit-learn.org/stable/modules/cross_validation.html), [Model evaluation](http://scikit-learn.org/stable/modules/model_evaluation.html)
- scikit-learn issue on GitHub: [MSE is negative when returned by cross_val_score](https://github.com/scikit-learn/scikit-learn/issues/2439)
- Section 5.1 of [An Introduction to Statistical Learning](http://www-bcf.usc.edu/~gareth/ISL/) (11 pages) and related videos: [K-fold and leave-one-out cross-validation](https://www.youtube.com/watch?v=nZAM5OXrktY&list=PL5-da3qGB5IA6E6ZNXu7dp89_uv8yocmf) (14 minutes), [Cross-validation the right and wrong ways](https://www.youtube.com/watch?v=S06JpVoNaA0&list=PL5-da3qGB5IA6E6ZNXu7dp89_uv8yocmf) (10 minutes)
- Scott Fortmann-Roe: [Accurately Measuring Model Prediction Error](http://scott.fortmann-roe.com/docs/MeasuringError.html)
- Machine Learning Mastery: [An Introduction to Feature Selection](http://machinelearningmastery.com/an-introduction-to-feature-selection/)
- Harvard CS109: [Cross-Validation: The Right and Wrong Way](https://github.com/cs109/content/blob/master/lec_10_cross_val.ipynb)
- Journal of Cheminformatics: [Cross-validation pitfalls when selecting and assessing regression and classification models](http://www.jcheminf.com/content/pdf/1758-2946-6-10.pdf)